tools/splitblocks/main.go

```diff
@@ -0,0 +1,232 @@
+	f.StringVar(&c.outputDir, "output.dir", "", "The output directory where split blocks will be written")
+	f.IntVar(&c.blockConcurrency, "block-concurrency", 5, "How many blocks can be split at once")
+	f.BoolVar(&c.dryRun, "dry-run", false, "If set blocks are not downloaded (except metadata) and splits are not performed; only what would happen is logged")
+	f.DurationVar(&c.maxBlockDuration, "max-block-duration", 24*time.Hour, "Max block duration, blocks larger than this or crossing a duration boundary are split.")
+}
+
+func (c *config) validate() error {
+	for _, blockID := range c.blocks {
+		if _, err := ulid.Parse(blockID); err != nil {
+			return errors.Wrap(err, "blocks contained an invalid block ID")
+		}
+	}
+	if c.maxBlockDuration < 2*time.Hour {
+		return fmt.Errorf("max-block-duration must be at least 2 hours")
+	}
+	if c.maxBlockDuration.Truncate(time.Hour) != c.maxBlockDuration {
+		return fmt.Errorf("max-block-duration should be aligned to hours")
+	}
+	if 24*time.Hour%c.maxBlockDuration.Truncate(time.Hour) != 0 {
+		return fmt.Errorf("max-block-duration should divide 24h without remainder")
+	}
+	if c.outputDir == "" {
+		return fmt.Errorf("output-dir is required")
+	}
+	if c.blockConcurrency < 1 {
+		return fmt.Errorf("block-concurrency must be positive")
+	}
+	return nil
+}
+
+func main() {
+	// Clean up all flags registered via init() methods of 3rd-party libraries.
+	flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	cfg := config{}
+	cfg.registerFlags(flag.CommandLine)
+
+	// Parse CLI arguments.
+	if err := flagext.ParseFlagsWithoutArguments(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	if err := cfg.validate(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	logger := log.NewLogfmtLogger(os.Stdout)
+	logger = log.With(logger, "ts", log.DefaultTimestampUTC)
+
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	if err := splitBlocks(ctx, cfg, logger); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+}
+
+func splitBlocks(ctx context.Context, cfg config, logger log.Logger) error {
+	bkt, err := bucket.NewClient(ctx, cfg.bucket, "bucket", logger, nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to create bucket")
+	}
+
+	// Helpful when not specifying -backend=filesystem (since it can use -filesystem.dir)
+	if cfg.bucketPrefix != "" {
+		bkt = bucket.NewPrefixedBucketClient(bkt, cfg.bucketPrefix)
+	}
+
+	var blocks []string
+	if len(cfg.blocks) > 0 {
+		blocks = cfg.blocks
+	} else {
+		blocks, err = listBlocks(ctx, bkt)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to list blocks", "err", err)
+			return errors.Wrapf(err, "failed to list blocks")
+		}
+	}
+
+	return concurrency.ForEachUser(ctx, blocks, cfg.blockConcurrency, func(ctx context.Context, blockString string) error {
```

Comment:
`ForEachUser` will continue with more blocks in case of errors. Is that what we want?

Reply:
No, I originally based the design off `copyblocks` so this was an oversight carried over from that. I changed it to `ForEachJob` to fail on the first error.